GameEngine/Agents/GreedyBestFirstAgent.cs

```diff
@@ -0,0 +1,51 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+
+namespace GameEngine.Agents
+{
+    public class GreedyBestFirstAgent : IAgent
+    {
+        public Queue<SearchNode> solution;
+
+        public Play FormulatePlay(GameState state)
+        {
+            solution = solution ?? GreedyBestFirstSearch(state);
+            var nextNode = solution.Dequeue();
+            if (nextNode.Parent.State != state)
+            {
+                throw new StateMismatchException();
+            }
+            return nextNode.Action;
+        }
+
+        private Queue<SearchNode> GreedyBestFirstSearch(GameState state)
+        {
+            var frontierNodes = new List<SearchNode> { new RootNode(state) };
+            while (true)
+            {
+                if (frontierNodes.Count == 0)
+                {
+                    throw new NoSolutionFoundException();
+                }
+                var node = frontierNodes.RemoveMin(H);
+                if ( H(node) == 0 )
+                {
+                    return new Queue<SearchNode>(node.Solution());
+                }
+                frontierNodes.AddRange(node.Expand());
+            }
+        }
+
+        private int H(SearchNode node)
+        {
+            return H(node.State);
+        }
+
+        public static int H(GameState state)
```

Comment:
❓ It was my understanding H should be an estimate of remaining path cost?

Reply:
In games, the heuristic is often just some function capable of inducing an order over game states.  h is generally short hand for cost-to-go, but only because most domains aren't 0-sum games.

💡 This is a reasonable heuristic, but I think it needs a more descriptive name.

Reply:
> 💡 This is a reasonable heuristic, but I think it needs a more descriptive name.

How do we feel about `TotalDistanceToNestAllOwls` vs `SumOfAllOwlsDistanceToNest` vs `WorstCaseNumberOfPlaysToGo`? Something else?

@rpherbig , this is a worst-case estimate of the remaining path cost. If a player were able, and played only cards that advanced an owl by a single space, this heuristic would represent the number of plays remaining at any given time. This means it (basically) always overestimates the real cost of the optimal solution. On the other hand, it doesn't ever overestimate the cost of the worst case solution. As the worst case gets less bad, we must be making good progress (so the thinking goes).

Incidentally, the overestimation of the optimal solution's cost means this heuristic is unsuited for A* since it's not [admissible](https://en.wikipedia.org/wiki/Admissible_heuristic). But as Bass says, it should still be "reasonable" in this case.

Reply:
I chose `WorstCaseNumberOfPlaysToGo` for this implementation of h(n), since that reflects the (probably) most interesting property of the heuristic.